tools/run_integration_test.py

```diff
@@ -0,0 +1,315 @@
+                        default=[],
+                        help='Exclude plugins from test.')
+    parser.add_argument(
+        '--run-on-changed-plugins',
+        default=False,
+        action='store_true',
+        help='Run the test on changed plugins. If --plugins is specified, \
+            this flag is ignored.')
+    parser.add_argument(
+        '--base-sha',
+        type=str,
+        default='',
+        help='The base sha used to determine git diff. This is useful when \
+            --run-on-changed-packages is specified. If not specified, \
+                merge-base is used as base sha.')
+    parser.add_argument(
+        '--timeout',
+        type=int,
+        default=120,
+        help='Timeout limit of each integration test in seconds. \
+            Default is 120 seconds.')
+
+    return parser.parse_args(args)
+
+
+def run_integration_test(plugin_dir, timeout):
+    """Runs integration test in the example package for plugin_dir
+
+    Currently the tools assumes that there's only one example package per plugin.
+
+    Args:
+        plugin_dir (str): The path to a single plugin directory.
+        timeout (int): Time limit in seconds before cancelling the test.
+
+    Returns:
+        PluginResult: The result of the plugin integration test.
+    """
+    example_dir = os.path.join(plugin_dir, 'example')
+    if not os.path.isdir(example_dir):
+        return PluginResult.fail([
+            'Missing example directory (use --exclude if this is intentional).'
+        ])
+
+    pubspec_path = os.path.join(example_dir, 'pubspec')
+    if not os.path.isfile(f'{pubspec_path}.yaml') and not os.path.isfile(
+            f'{pubspec_path}.yml'):
+        # TODO: Support multiple example packages.
+        return PluginResult.fail(['Missing pubspec file in example directory'])
+
+    integration_test_dir = os.path.join(example_dir, 'integration_test')
+    if not os.path.isdir(integration_test_dir) or not os.listdir(
+            integration_test_dir):
+        return PluginResult.fail([
+            'Missing integration tests (use --exclude if this is intentional).'
+        ])
+
+    errors = []
+    completed_process = subprocess.run('flutter-tizen pub get',
+                                       shell=True,
+                                       cwd=example_dir,
+                                       stderr=subprocess.PIPE,
+                                       stdout=subprocess.PIPE)
+    if completed_process.returncode != 0:
+        if not completed_process.stderr:
+            errors.append('pub get failed. Make sure the pubspec file \
+                    in your project is valid.')
+        else:
+            errors.append(completed_process.stderr)
+        return PluginResult.fail(errors)
+
+    # This prevents ci test failures when depending packages accidentally
+    # publishes breaking changes as minor or patch versions.
+    # (e.g. package_info_plus 1.0.4 breaks PackageInfo's public constructor.)
```

Comment:
Shouldn't this error be detected by the test tool?

Reply:
Good point, yes I think it should be detected and we should fix the problem when such error occurs. According to https://semver.org/#what-do-i-do-if-i-accidentally-release-a-backwards-incompatible-change-as-a-minor-version, `package_info_plus` should release a new patch version that restores backward compatibility. I've reported to the maintainers: https://github.com/fluttercommunity/plus_plugins/issues/392.

Maybe I could add an option like "--run-on-downgraded-packages", we might encounter similar problems in the future once the CI has been setup.

Reply:
To be more specific, the following changes to platform interface classes will break existing Tizen implementations (FFI-based packages only).

- Adding a property without providing a default implementation
- Changing an existing methods signature (such as adding an optional parameter)

What we have to do in such situation is to update our implementations as soon as possible. The errors are not false alarm, as the users will encounter exactly the same error when depending on our packages. Expecting the users to run `flutter pub downgrade` by themselves doesn't make sense because they will not even understand the situation.

Reply:
Let me see if I've got it right.

If the package we depend on(let's say A) make those mentioned changes, normally it will be released after increasing the major version, for example, 1.0.0 to 2.0.0. At this point if our package(let's say B)'s pubspec was:

```
dependency:
    A: ^1.0.0
```

B will still build normally because pub tool doesn't automatically get the next higher major version. 

But if A was accidentally released as the next minor version, for example, 1.0.0 to 1.1.0, the pub tool will get A-1.1.0 and the build will fail due to breaking changes.

My initial thought was to wait for A to fix the problem by releasing a new minor version that removes all breaking changes and then release a new major version (what the maintainers probably wanted to do in the first place). And In the meantime allow the CI to pass with the minimum stable(?) version until A is fixed.

Instead in such cases, we should let the CI fail and resolve the breaking changes immediately because:

> as the users will encounter exactly the same error when depending on our packages. Expecting the users to run flutter pub downgrade by themselves doesn't make sense because they will not even understand the situation.

Is this correct?

Reply:
> My initial thought was to wait for A to fix the problem by releasing a new minor version that removes all breaking changes and then release a new major version

Changing a method signature or adding a property is a reasonably "minor" change (although it's against backwards compatibility), and the maintainers of platform interface packages do not increase the major version in such situation. It's not mandatory for the maintainers to follow the versioning policy strictly, and we can't force them to follow it.

So yes, if any breaking change has (unfortunately) been made to a platform interface class, we need to resolve it immediately.

Reply:
Ah okay, I got it now, thanks for the detailed explanation. I'll remove the "downgrading" code and resolve `package_info_plus` in a separate PR.